plugin/v2/plugin.go

```
@@ -0,0 +1,162 @@
+)
+
+// Store the last known primary key version resource name
+// to return as KeyId in case when the Cloud KMS service is not reachable
+// because KeyId in StatusResponse cannot be empty and shouldn't trigger key migration
+// in case of transient remote service unavailability
+var lastKeyId string
+
+// Regex to extract Cloud KMS key resource name from the key version resource name
+var keyResourceRegEx = regexp.MustCompile(`projects\/[^/]+\/locations\/[^/]+\/keyRings\/[^/]+\/cryptoKeys\/[^/:]+`)
+
+type Plugin struct {
+	*plugin.AbstractPlugin
+}
+
+// New constructs Plugin.
+func NewPlugin(keyService *cloudkms.ProjectsLocationsKeyRingsCryptoKeysService, keyURI, keySuffix string, pathToUnixSocketFile string) *Plugin {
+	lastKeyId = getKeyId(keyURI, keySuffix)
+	p := &Plugin{
+		AbstractPlugin: &plugin.AbstractPlugin{
+			KeyService:       keyService,
+			KeyURI:           keyURI,
+			KeySuffix:        keySuffix,
+			PathToUnixSocket: pathToUnixSocketFile,
+		},
+	}
+	p.AbstractPlugin.Plugin = p
+	return p
+}
+
+// Status returns the version of KMS API version that plugin supports.
+// Response also contains the status of the plugin, which is calculated as availability of the
+// encryption key that the plugin is confinged with, and the current primary key version.
+// kube-apiserver will provide this key version in Encrypt and Decrypt calls and will be able
+// to know whether the remote CLoud KMS key has been rotated or not.
+func (g *Plugin) Status(ctx context.Context, request *StatusRequest) (*StatusResponse, error) {
+	defer plugin.RecordCloudKMSOperation("encrypt", time.Now().UTC())
+
+	var response StatusResponse
+	req := &cloudkms.EncryptRequest{Plaintext: ping}
+	resp, err := g.KeyService.Encrypt(g.KeyURI, req).Context(ctx).Do()
+	if err != nil {
+		plugin.CloudKMSOperationalFailuresTotal.WithLabelValues("encrypt").Inc()
+		response = StatusResponse{Version: apiVersion, Healthz: keyNotReachable, KeyId: lastKeyId}
```

Why not return the original error here?

---

The period for calling the Status method determines Kubernetes.
The API server polls the Status procedure call approximately every minute when everything is healthy, and every 10 seconds when the plugin is not healthy. 
As I learned from Di Jin, that is what the GKE implementation doing. Using paid encrypt/decrypt call pair instead of a free get call has advantages: 1) no additional IAM permissions required 2) the status check is very accurate.
Calling Cloud KMS API twie in a minute should not be a big overhead.